kloppy/domain/services/event_deducers/carry.py

```diff
@@ -0,0 +1,142 @@
+import bisect
+import uuid
+from datetime import timedelta
+
+from kloppy.domain import (
+    EventDataset,
+    EventType,
+    BodyPart,
+    CarryResult,
+    GenericEvent,
+    EventFactory,
+    Unit,
+)
+from kloppy.domain.services.event_deducers.event_deducer import (
+    EventDatasetDeduducer,
+)
+
+
+class CarryDeducer(EventDatasetDeduducer):
+    min_carry_length_meters = 3
+    max_carry_length_meters = 60
+    max_carry_duration = timedelta(seconds=10)
+    event_factory = EventFactory()
+
+    def deduce(self, dataset: EventDataset):
+        pitch = dataset.metadata.pitch_dimensions
+
+        new_carries = []
+
+        valid_event_types = [
+            EventType.PASS,
+            EventType.SHOT,
+            EventType.TAKE_ON,
+            EventType.CLEARANCE,
+            EventType.INTERCEPTION,
+            EventType.DUEL,
+            EventType.RECOVERY,
+            EventType.MISCONTROL,
+            EventType.GOALKEEPER,
+        ]
+
+        for idx, event in enumerate(dataset.events):
+            if event.event_type not in valid_event_types:
+                continue
+            idx_plus = 1
+            generic_next_event = True
+            while idx + idx_plus < len(dataset.events) and generic_next_event:
+                next_event = dataset.events[idx + idx_plus]
+
+                if next_event.event_type in [
+                    EventType.GENERIC,
+                    EventType.PRESSURE,
+                ]:
+                    idx_plus += 1
+                    continue
+                else:
+                    generic_next_event = False
+                if not event.team.team_id == next_event.team.team_id:
+                    continue
+
+                if next_event.event_type not in valid_event_types:
+                    continue
+                # not headed shot
+                if (
+                    (hasattr(next_event, "body_part"))
```

Comment:
An `Event` does never have a "body_part" attribute. The body part is added as a qualifier.